record/envelope.go

```diff
@@ -0,0 +1,302 @@
+package record
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sync"
+
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/record/pb"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/multiformats/go-varint"
+)
+
+// Envelope contains an arbitrary []byte payload, signed by a libp2p peer.
+//
+// Envelopes are signed in the context of a particular "domain", which is a
+// string specified when creating and verifying the envelope. You must know the
+// domain string used to produce the envelope in order to verify the signature
+// and access the payload.
+type Envelope struct {
+	// The public key that can be used to verify the signature and derive the peer id of the signer.
+	PublicKey crypto.PubKey
+
+	// A binary identifier that indicates what kind of data is contained in the payload.
+	// TODO(yusef): enforce multicodec prefix
+	PayloadType []byte
+
+	// The envelope payload.
+	RawPayload []byte
+
+	// The signature of the domain string :: type hint :: payload.
+	signature []byte
+
+	// the unmarshalled payload as a Record, cached on first access via the Record accessor method
+	cached         Record
+	unmarshalError error
+	unmarshalOnce  sync.Once
+}
+
+var ErrEmptyDomain = errors.New("envelope domain must not be empty")
+var ErrEmptyPayloadType = errors.New("payloadType must not be empty")
+var ErrInvalidSignature = errors.New("invalid signature or incorrect domain")
+
+// MakeEnvelope constructs a new Envelope using the given privateKey.
+//
+// The required 'domain' string contextualizes the envelope for a particular purpose,
+// and must be supplied when verifying the signature.
+//
+// The 'PayloadType' field indicates what kind of data is contained and must be non-empty.
+func MakeEnvelope(privateKey crypto.PrivKey, domain string, payloadType []byte, payload []byte) (*Envelope, error) {
```

Comment:
Both MakeEnvelope methods can be merged if we always pass in an instance of the `Record` interface (once it's been enhanced with the `Codec()` and `Domain()` methods).